pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
+            logger.debug(
+                "DeviceDerivatives created with (%s, %s, %s)",
+                device,
+                execution_config,
+                gradient_kwargs,
+            )
+
+        self._device = device
+        self._execution_config = execution_config
+        self._gradient_kwargs = gradient_kwargs
+
+        self._uses_new_device = not isinstance(device, qml.Device)
+
+        # only really need to keep most recent entry, but keeping 10 around just in case
+        self._results_cache = LRUCache(maxsize=10)
+        self._jacs_cache = LRUCache(maxsize=10)
+
+    def _dev_execute_and_compute_derivatives(self, tapes: Batch):
+        """
+        Converts tapes to numpy before computing the the results and derivatives on the device.
+
+        Dispatches between the two different device interfaces.
+        """
+        numpy_tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
+        if self._uses_new_device:
+            return self._device.execute_and_compute_derivatives(numpy_tapes, self._execution_config)
+        return self._device.execute_and_gradients(numpy_tapes, **self._gradient_kwargs)
+
+    def _dev_execute(self, tapes: Batch):
+        """
+        Converts tapes to numpy before computing just the results on the device.
+
+        Dispatches between the two different device interfaces.
+        """
+        numpy_tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
+        if self._uses_new_device:
+            return self._device.execute(numpy_tapes, self._execution_config)
+        return self._device.batch_execute(numpy_tapes)
+
+    def _dev_compute_derivatives(self, tapes: Batch):
+        """
+        Converts tapes to numpy before computing the derivatives on the device.
+
+        Dispatches between the two different device interfaces.
+        """
+        numpy_tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
+        if self._uses_new_device:
+            return self._device.compute_derivatives(numpy_tapes, self._execution_config)
+        return self._device.gradients(numpy_tapes, **self._gradient_kwargs)
+
+    def execute(self, tapes: Batch):
+        """Forward pass used to cache the results and jacobians.
+
+        Args:
+            tapes (tuple[`~.QuantumScript`]): the batch of tapes to execute and take derivatives of
+
+        Returns:
+            ResultBatch: the results of the execution.
+
+        Side Effects:
+            Caches both the results and jacobian into ``_results_cache`` and ``_jacs_cache``.
+
+        """
+        if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+            logger.debug("Forward pass called with %s", tapes)
+        results, jac = self._dev_execute_and_compute_derivatives(tapes)
+        self._results_cache[tapes] = results
+        self._jacs_cache[tapes] = jac
+        return results
+
+    def execute_and_compute_jvp(self, tapes: Batch, tangents):
```

Comment:
Are we not missing `execute_and_compute_vjp`?

Reply:
No interface will bind `execute_and_compute_vjp`, as you can't really have the `dy` until you perform the forward pass. If in the future we need it for a situation, then we can add that method in.

Reply:
Thanks!